### DISCUSSION TITLE
Questions about PR workflows and actions/checkout@v2

### DISCUSSION THREAD

**Discussion by harluss** [community/community#25961] [Actions]
<p>Hi,</p>
<p>I have 2 questions about GitHub Actions:</p>
<ol>
<li>Could someone please explain to me how <code>actions/checkout@v2</code> works for pull requests as this seem unclear to me. If my workflow <code>yml</code> looks like this:</li>
</ol>
<pre><code class="lang-auto">on:
  pull_request:
    branches: [master]

...

    steps:
      - name: Checkout repository
        uses: actions/checkout@v2
        with:
          ref: ${{ github.event.pull_request.head.sha }}
</code></pre>
<p>Any pull request to master branch will checkout the branch that triggered the workflow and run the rest of the workflow (tests) on that branch.<br>
But, what is the default behaviour for PR if I leave it like this (skipping the <code>with: ref:</code> part):</p>
<pre><code class="lang-auto">steps:
      - name: Checkout repository
        uses: actions/checkout@v2
</code></pre>
<p>Will it checkout and run tests on the branch that triggered the PR as well? If so, is there any reason/adventage to being more specific and adding <code>with: ref:</code>?</p>
<ol start="2">
<li>Is there any particular way of limiting branches that can create a PR to protected branch? Let’s say to allow <code>Dev -&gt; Master</code>, but not <code>random branch -&gt; Master</code>?</li>
</ol>

**Comment by Yanjingzhu** [comment, marked as answer]
<ol>
<li>The default ref of <a href="https://github.com/actions/checkout#usage" rel="nofollow noopener">actions/checkout@v2</a> is to the reference or SHA for that event. The GITHUB_REF of pull request event is the PR merge branch. So if you doesn’t specify with ref part, it checks out the merged source code.<br>
<code> github.event.pull_request.head.sha</code> points to the pushed commit of the pull request, it is on the PR source branch.</li>
<li>As far as I know, there is no way to prevent a specific branch merge to master.</li>
</ol>

**Comment by harluss** [comment]
<aside class="quote group-github-partners" data-username="yanjingzhu" data-post="2" data-topic="122347">
<div class="title">
<div class="quote-controls"></div>
 yanjingzhu:</div>
<blockquote>
<p>The GITHUB_REF of pull request event is the PR merge branch. So if you doesn’t specify with ref part, it checks out the merged source code.</p>
</blockquote>
</aside>
<p>Oh, so it creates what would be the final outcome of the PR after approval and runs the tests (or whatever else is specified in the workflow) against that?</p>
<p>Fantastic, thank you for the clarification.</p>

**Comment by Yanjingzhu** [comment]
<p>Yeah, using the default ref matches the normal scenario to build PR merged code.<br>
I forget to attach this screenshot in my above comment. Add it here:<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/e534ebbe917206ff7ed13b61ff83c1477e3d4424.png" data-download-href="/uploads/short-url/wHEIPVX2EBEVnLYMyDNJxyMFuio.png?dl=1" title="PR ref" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181094313-7e25143f-1569-417a-acb7-5aa16461e5e1.png" alt="PR ref" data-base62-sha1="wHEIPVX2EBEVnLYMyDNJxyMFuio" width="545" height="500" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/e534ebbe917206ff7ed13b61ff83c1477e3d4424_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">PR ref</span><span class="informations">769×705 22 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by kanjurer** [comment]
You can prevent other branches to merge into Master by setting 
```
ref: ${{ github.event.pull_request.head.ref }}
```
